stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +229,186 @@ class task_system {
+            if (_done) return;
         }
     }
 
 public:
-    task_system() {
+    priority_task_system() {
+        for (auto& q : _q) {
+            std::vector<notification_queue> queues{_count};
+            std::swap(q, queues);
+            for (unsigned n = 0; n != _count; ++n) {
+                q[n].set_context(_thread_contexts[n]._ready);
+            }
+        }
         for (unsigned n = 0; n != _count; ++n) {
-            _threads.emplace_back([&, n] { run(n); });
+            _thread_contexts[n]._thread = std::thread([&, n] { run(n); });
         }
     }
 
-    ~task_system() {
-        for (auto& e : _q)
-            e.done();
+    ~priority_task_system() {
+        _done = true;
+        for (auto& context : _thread_contexts)
+            context._ready.notify_all();
 
-        for (auto& e : _threads)
-            e.join();
+        for (auto& context : _thread_contexts)
+            context._thread.join();
     }
 
-    template <typename F>
-    void operator()(F&& f) {
+    template <std::size_t P, typename F>
+    void execute(F&& f) {
+        static_assert(P < 3, "More than 3 priorities are not known!");
         auto i = _index++;
 
-        for (unsigned n = 0; n != _count; ++n) {
-            if (_q[(i + n) % _count].try_push(std::forward<F>(f))) return;
+        for (unsigned n = 0; n != _spin / _count; ++n) {
+            if (_q[P][(i + n) % _count].try_push(std::forward<F>(f))) return;
         }
 
-        _q[i % _count].push(std::forward<F>(f));
+        _q[P][i % _count].push(std::forward<F>(f));
+    }
+};
+
+inline priority_task_system& pts() {
+    static priority_task_system only_task_system;
+    return only_task_system;
+}
+
+template <executor_priority P = executor_priority::medium>
+struct task_system
+{
+    using result_type = void;
+
+    void operator()(task<void()> f) const {
+        pts().execute<static_cast<std::size_t>(P)>(std::move(f));
     }
 };
 
+
 #endif
 
 #if (STLAB_TASK_SYSTEM == STLAB_TASK_SYSTEM_PORTABLE) || \
     (STLAB_TASK_SYSTEM == STLAB_TASK_SYSTEM_WINDOWS)
 
-struct default_executor_type {
+template <executor_priority P = executor_priority::medium>
+struct executor_type {
     using result_type = void;
 
     void operator()(task<void()> f) const {
-        static task_system only_task_system;
+        static task_system<P> only_task_system;
         only_task_system(std::move(f));
     }
 };
 
 #endif
+
 /**************************************************************************************************/
 
 } // namespace detail
 
 /**************************************************************************************************/
 
-constexpr auto default_executor = detail::default_executor_type{};
+constexpr auto low_executor = detail::executor_type<detail::executor_priority::low>{};
+constexpr auto default_executor = detail::executor_type<detail::executor_priority::medium>{};
+constexpr auto high_executor = detail::executor_type<detail::executor_priority::high>{};
```

Comment:
Since all of these executors are instances of the same tasking system there is nothing that really makes one a higher/lower priority than another. I am guessing since this is the portable version we won't want to mess with platform specific thread priorities and just leave it as a polly fill. Is that correct?

Reply:
It is correct, that they work all on the same task_system. But the different priorities are taken into account by this tasking system. You are correct. All thread shall have the same prio and just the tasks are prioritized on the system according to their priority.

Reply:
I am only slightly familiar with how the underlying task system functions for each platform. However from a users perspective on wanting to write portable and consistent code it might yield unexpected results. I am not saying anything has to change but it is interesting to think about.

On the portable task system the labels of `low`, `medium`, and `high` are simply organizational and have no bearing on priority. However there can be an implicit priority that is applied by the origination of the user but it may not be what the user expects.

For example if you add 50 tasks to the `high` executor, 20 to the `medium`, and 10 to the `low` what would the user expect to happen (assuming all tasks are equivalent)? Naturally the user would think that the `high` executor tasks would finish before the `medium` and the `medium` before the `low` but in fact the exact opposite would happen.

Perhaps overall a better way to organize tasks with executors would be through bounds like `cpu`, `io`, and `blocking`.

I think executors like `high`, `medium`, and `low` would only be useful for `cpu` bound tasks where the underlying thread priority matches the executor and/or task stealing occurs with affinity based on the priority.

Thoughts? @FelixPetriconi @fosterbrereton

Reply:
High, default and low _have_ an influence. Your assumption on the execution of tasks is not completely correct. The high tasks does not have to be finished, before the other tasks start. The default and low prio tasks will start after all (or most of the high) prio tasks are started.

Three queues core each are associated. One high, default and low. The run function then first searches through all high queues for tasks - here index yero - , if there is one, it will pick it and execute it. If there is no high prio tasks, then it searches through the medium prio queues and so on.

There is a test that prints the ratio of the correctness of distribution. The measurement is not 100% correct, but a an approximation.

Reply:
You are absolutely correct! I was thinking about how the old `task_system` worked and kinda merged the two versions in my head. I missed that all the `task_system`s now reference the same `priority_task_system` and iterate over the queues. Don't mind me.